### PR TITLE
Always use RevealAll for const eval queries

### PR DESCRIPTION
**Pull request by oli-obk**
implements what is described in https://github.com/rust-lang/rust/pull/116803#discussion_r1364089471

Using `UserFacing` for const eval does not make sense anymore, unless we significantly change things like avoiding revealing opaque types.

New tests are copied from https://github.com/rust-lang/rust/pull/101478

### CODE REVIEW

**Comment by oli-obk** [comment]
yay free const eval performance improvement.

File: compiler/rustc_const_eval/src/const_eval/eval_queries.rs [hunk 1]
```diff
@@ -226,16 +226,7 @@ pub fn eval_to_const_value_raw_provider<'tcx>(
     key: ty::ParamEnvAnd<'tcx, GlobalId<'tcx>>,
 ) -> ::rustc_middle::mir::interpret::EvalToConstValueResult<'tcx> {
     // see comment in eval_to_allocation_raw_provider for what we're doing here
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/const_eval/eval_queries.rs:228]: That comment now probably needs updating?

The assertion error or the comment should explain what a rustc dev should do when they hit this panic.

**Review by RalfJung — COMMENTED** [review]
This removes a bunch of code that I never understood and improves performance (except, strangely, in one benchmark). I'm all for it :)

File: tests/ui/specialization/ctfe/default-assoc-const.rs [hunk 2]
```diff
@@ -0,0 +1,21 @@
+// ICE fixed by #119821.
+//
+// See that PR for more details.
+//
+// check-pass
```
Comment (lcnr) [inline tests/ui/specialization/ctfe/default-assoc-const.rs:5]: doesn't this pass rn? https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=7dba8d9aa1ced839c563e298a595e431

File: tests/ui/specialization/ctfe/default-assoc-type.rs [hunk 3]
```diff
@@ -0,0 +1,29 @@
+// ICE fixed by #119821.
```
Comment (lcnr) [inline tests/ui/specialization/ctfe/default-assoc-type.rs:1]: same here https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=a01dd24080ed88b3a4249714dcb955c0

Comment (oli-obk) [reply tests/ui/specialization/ctfe/default-assoc-type.rs:1]: also still passes after this PR

File: tests/ui/specialization/ctfe/default-assoc-const.rs (hunk 2 above)
Comment (lcnr) [reply tests/ui/specialization/ctfe/default-assoc-const.rs:5]: > // ICE fixed by #119821.

that's why I commented this. It did not ICE before :sweat_smile:

Comment (oli-obk) [reply tests/ui/specialization/ctfe/default-assoc-const.rs:5]: Whoops, I updated the other tests' comments, but forgot this one

**Comment by oli-obk** [comment]
fixed test comments

File: tests/ui/impl-trait/in-ctfe/match-arm-exhaustive.rs [hunk 4]
```diff
@@ -0,0 +1,24 @@
+//! Check that pattern matching can observe the hidden type of opaqaue types.
```
Comment (lqd) [inline tests/ui/impl-trait/in-ctfe/match-arm-exhaustive.rs:1]: (Tiny typo happening twice)

```suggestion
//! Check that pattern matching can observe the hidden type of opaque types.
```

**Review by lcnr — COMMENTED** [review]
please update comments and add this test :grin: r=me after that
```rust
trait MyTrait: Copy {
    const ASSOC: usize;
}

impl MyTrait for u8 {
    const ASSOC: usize = 32;
}

const fn yeet() -> impl MyTrait {
    0u8
}

const fn output<T: MyTrait>(_: T) -> usize {
    <T as MyTrait>::ASSOC
}

struct Foo<'a>(&'a ())
const NEED_REVEAL_ALL: usize = output(yeet());

fn promote_div() -> &'static usize {
    &(10 / NEED_REVEAL_ALL)
}
fn main() {}
```

**Pull request merged**